### ISSUE TITLE
Exceptions won't log

### ISSUE THREAD

**Issue by moshest** [lazywithclass/winston-cloudwatch#44]
I notice that when my app crashes, the logs won't set to Amazon.
I think it's because this plugin send the callback immediately and not wait for the sending interval.

My configuration include `{handleExceptions: true}`

**Comment by lazywithclass** [comment]
@moshest I'm not sure I understood your problem; if the process running node crashes winston-cloudwatch will no longer be running too, as it is running in that same process.

**Comment by moshest** [comment]
Winston has the ability to log crashes and then exit. Please see here: https://github.com/winstonjs/winston#handling-uncaught-exceptions-with-winston

But because the logger [call the callback immediately](https://github.com/lazywithclass/winston-cloudwatch/blob/master/index.js#L60) the process exit before the logging is complete.

**Comment by lazywithclass** [comment]
Yeah I'm reading that first link right now. Could you please paste here how you're using it?

Thanks for the pointer in the code I think I get it now, I will have to code a new feature, it doesn't look too difficult.

**Comment by moshest** [comment]
`handleExceptions` make a bind to [`'uncaughtException'` event](https://nodejs.org/api/process.html#process_event_uncaughtexception), You can see it [here](https://github.com/winstonjs/winston/blob/de69bccf957cca2247ef529c2b66c88059adbae6/lib/winston/logger.js#L440).

You may want to extend the method [`logException`](https://github.com/winstonjs/winston/blob/d4fdbadc2f4ab8408261497a116ef80e0f9475a0/lib/winston/transports/transport.js#L115) on your `Transport` and handle it correctly.

**Comment by lazywithclass** [comment]
Yes I meant an example on how you are using it, like what you are passing to `winston.handleExceptions`.

With that example it will be easier for me to test.

**Comment by moshest** [comment]
```js
const winston = require('winston');
const WinstonCloudWatch = require('winston-cloudwatch');
const transports = [];

// more transports here..

transports.push(new WinstonCloudWatch({
  level: 'info',
  handleExceptions: true,
  logGroupName: '...',
  logStreamName: '...',
  awsAccessKeyId: '...',
  awsSecretKey: '...',
  awsRegion: '...'
}));

const logger = new winston.Logger({
  transports,
});
```

**Comment by lazywithclass** [comment]
Ah ok I see, you enable it as a flag. Thanks!

**Comment by lazywithclass** [comment]
Hey @moshest I had some free time and looked into this.

I still have a doubt about how to implement this and you might help me clear it, in winston-cloudwatch there's a `setInterval` that sends logs to AWS, I'm not sending them as soon as you call `log`, so there's no real point in which I can say "I am done call the `callback`".

I am writing a small test to reproduce a crash and see what happens.

**Comment by moshest** [comment]
In my opinion, you should stack all the callbacks and simply call them once the `setInterval` accord.

**Comment by lazywithclass** [comment]
Mh... that could be one way of doing it, nice.
I still want to see the problem in action though, before coding my solution.
Will keep you updated.

**Comment by moshest** [comment]
Any news?

**Comment by lazywithclass** [comment]
I am sorry it's taking this much time, I am having some issue at work in this period.
What's slowing this down is the proof of concept test that simulates your scenario, the change per se it's easy to do.

**Comment by lazywithclass** [comment]
Hey @moshest, I now have more time to look into this.
I've set up a simple scenario to reproduce your problem:

```JS
var winston = require('winston'),
WinstonCloudWatch = require('../index');

winston.add(WinstonCloudWatch, {
  logGroupName: 'testing',
  logStreamName: 'first',
  handleExceptions: true
});

setTimeout(function() {
  winston.error('winston error');
}, 0);

var oops = {};
oops.byebye();
```

Your process is going to exit regardless of me waiting for `WinstonCloudWatch.prototype.add` to finish with a callback, this is because of how uncaught exceptions are handled by Winston: https://github.com/winstonjs/winston#to-exit-or-not-to-exit. 
Your process will exit after 3000ms if you don't set `exitOnError` as `false` (the default value is `true`).

I'm happy to help you with any problem you're having with my library, please do continue replying if the problem still occurs after this change, I'm leaving the issue open.

**Comment by moshest** [comment]
We just need to make sure that the exception on this example will logged correctly.

```js
const winston = require('winston');
const WinstonCloudWatch = require('../index');

winston.add(WinstonCloudWatch, {
  logGroupName: 'testing',
  logStreamName: 'first',
  handleExceptions: true
});

// this error should log immediately, before the 3000ms timeout.
throw new Error('test me');
```

You can find out if an error triggered from an exception by testing if the error message start with the string `"uncaughtException: "`. You can see in on Winston [here](https://github.com/winstonjs/winston/blob/de69bccf957cca2247ef529c2b66c88059adbae6/lib/winston/logger.js#L649).

**Comment by lazywithclass** [comment]
Ok, I will set up a test CloudWatch env. Having some problems to login with my MFA at the moment.

**Comment by lazywithclass** [comment]
@moshest Ok so the problem is that, with the configuration you passed me, it takes more than the Winston timeout to successfully post a log record into CloudWatch.

You can check this by changing [this line](https://github.com/winstonjs/winston/blob/master/lib/winston/logger.js#L674) in your local Winston instance in `node_modules/` to, for example, `10000`, you will see the log being successfully added.

At this point there are two ways:

1. you make use of Winston's `exitOnError`
2. we propose a PR that lets us have a configurable waiting time, before calling `process.exit()`

**Comment by moshest** [comment]
I think that both options are bad.

1. It's a bad habit to ignore errors. Node should exit on exceptions.
2. It's seems reasonable that if there is an error, the process will exit in 3 seconds.

Please see my [pull request](https://github.com/lazywithclass/winston-cloudwatch/pull/47) fo a solution.

**Comment by lazywithclass** [comment]
Well... I think "bad" it's a point of view here, otherwise they wouldn't have added `exitOnError` in the first place. Especially because it really depends on the error, if it's something you can recover from you can definitely do that.
I'm confused why exactly 3 seconds seems reasonable :D 

But I am digressing, I will have a look at the PR, thank you.

**Comment by moshest** [comment]
It's bad because if on some error you DO want to exit, the logs wouldn't work.

Regarding the 3 seconds, If a program has an error and should be killed, 3 seconds is more then enough.

**Comment by lazywithclass** [comment]
I'm not used to estimate generic solutions, I prefer to decide based on the case at hand, so I'll pass on the discussion :D

**Comment by lazywithclass** [comment]
@moshest Available in 1.6.0

**Comment by moshest** [comment]
Thanks!
